File: src/wiredtiger/session.cc
#include <wiredtiger.h>
#include "helpers.h"
#include "cursor.h"
#include "../helpers.h"
#include "session.h"

#include <vector>

using namespace std;

namespace wiredtiger {
  WiredTigerSession::WiredTigerSession(WT_SESSION *session):
    session(session)
  {
    if (session->app_private != NULL) {
      printf("Binding session multiple times!\n");
      abort();
    }
    session->app_private = this;
  }

  WiredTigerSession::~WiredTigerSession() {
    printf("Reset: %p\n", this);
    // TODO: close session if not already
  }

  bool WiredTigerSession::isInTransaction() {
    return _isInTransaction;
  }

  int WiredTigerSession::create(char* typeAndName, char* specs) {
    return session->create(
      session,
      typeAndName,
      specs
    );
  }

  int WiredTigerSession::close(char* config) {
    return session->close(session, config);
  }
  int WiredTigerSession::join(Cursor* joinCursor, Cursor* refCursor, char* config) {
    return session->join(session, joinCursor->getWTCursor(), refCursor->getWTCursor(), config);
  }
  int WiredTigerSession::openCursor(char* cursorSpec, Cursor** cursor) {
    WT_CURSOR* wtCursor;
    RETURN_IF_ERROR(session->open_cursor(session, cursorSpec, NULL, NULL, &wtCursor));
    *cursor = new Cursor(wtCursor);
    return 0;
  }
  int WiredTigerSession::openCursor(char* cursorSpec, char* config, Cursor** cursor) {
    WT_CURSOR* wtCursor;
    RETURN_IF_ERROR(session->open_cursor(session, cursorSpec, NULL, config, &wtCursor));
    *cursor = new Cursor(wtCursor);
    return 0;
  }

  int WiredTigerSession::beginTransaction(char* config) {
    _isInTransaction = true;
    return session->begin_transaction(session, config);
  }
  int WiredTigerSession::prepareTransaction(char* config) {
    return session->prepare_transaction(session, config);
  }
  int WiredTigerSession::commitTransaction(char* config) {
    _isInTransaction = false;
    return session->commit_transaction(session, config);
  }
  int WiredTigerSession::rollbackTransaction(char* config) {
    _isInTransaction = false;
    return session->rollback_transaction(session, config);
  }
}

File: src/glue/session.rs

use crate::external::wiredtiger::*;

use super::cursor::InternalCursor;
use super::utils::{get_fn, unwrap_string_and_dealloc, unwrap_string_or_null};
use super::error::*;

pub struct InternalSession {
  session: *mut WT_SESSION
}

impl InternalSession {
  pub fn new(session: *mut WT_SESSION) -> Self {
    InternalSession { session }
  }

  pub fn get_session_ptr(&self) -> *mut WT_SESSION {
    return self.session;
  }

  fn get_session(&self) -> Result<WT_SESSION, GlueError> {
    if self.session == std::ptr::null_mut() {
      return Err(GlueError::for_glue(GlueErrorCode::UnlikelyNoPtr));
    }
    unsafe {
      Ok(*self.session)
    }
  }

  pub fn create(&self, uri: String, config: Option<String>) -> Result<(), GlueError> {
    let result = unsafe {
      get_fn(self.get_session()?.create)?(
        self.session,
        unwrap_string_and_dealloc(uri).as_ptr(),
        unwrap_string_or_null(config).as_ptr()
      )
    };

    if result == 0 {
      return Ok(());
    }
    Err(GlueError::for_wiredtiger( result))
  }

  pub fn close(&self, config: Option<String>) -> Result<(), GlueError> {
    let result = unsafe {
      get_fn(self.get_session()?.close)?(
        self.session,
        unwrap_string_or_null(config).as_ptr()
      )
    };

    if result == 0 {
      return Ok(());
    }
    Err(GlueError::for_wiredtiger( result))
  }

  pub fn open_cursor(&self, cursor_spec: String, config: Option<String>) -> Result<InternalCursor, GlueError> {
    let mut cursor: *mut WT_CURSOR = std::ptr::null_mut();
    let error: std::os::raw::c_int = unsafe { get_fn(self.get_session()?.open_cursor)?(
      self.session,
      unwrap_string_and_dealloc(cursor_spec).as_ptr(),
      std::ptr::null_mut(),
      unwrap_string_or_null(config).as_ptr(),
      &mut cursor
    ) };
    if error != 0 {
      return Err(GlueError::for_wiredtiger(error));
    }
    InternalCursor::new(cursor)
  }

  pub fn begin_transaction(&self, config: Option<String>) -> Result<(), GlueError> {
    let result = unsafe {
      get_fn(self.get_session()?.begin_transaction)?(
        self.session,
        unwrap_string_or_null(config).as_ptr()
      )
    };

    if result == 0 {
      return Ok(());
    }
    Err(GlueError::for_wiredtiger( result))
  }

  pub fn commit_transaction(&self, config: Option<String>) -> Result<(), GlueError> {
    let result = unsafe {
      get_fn(self.get_session()?.commit_transaction)?(
        self.session,
        unwrap_string_or_null(config).as_ptr()
      )
    };

    if result == 0 {
      return Ok(());
    }
    Err(GlueError::for_wiredtiger( result))
  }

  pub fn rollback_transaction(&self, config: Option<String>) -> Result<(), GlueError> {
    let result = unsafe {
      get_fn(self.get_session()?.rollback_transaction)?(
        self.session,
        unwrap_string_or_null(config).as_ptr()
      )
    };

    if result == 0 {
      return Ok(());
    }
    Err(GlueError::for_wiredtiger( result))
  }

  pub fn prepare_transaction(&self, config: Option<String>) -> Result<(), GlueError> {
    let result = unsafe {
      get_fn(self.get_session()?.prepare_transaction)?(
        self.session,
        unwrap_string_or_null(config).as_ptr()
      )
    };

    if result == 0 {
      return Ok(());
    }
    Err(GlueError::for_wiredtiger( result))
  }

  pub fn join(&self, join_cursor: &InternalCursor, ref_cursor: &InternalCursor, config: Option<String>) -> Result<(), GlueError> {
    let result = unsafe {
      get_fn(self.get_session()?.join)?(
        self.session,
        join_cursor.cursor,
        ref_cursor.cursor,
        unwrap_string_or_null(config).as_ptr()
      )
    };

    if result == 0 {
      return Ok(());
    }
    Err(GlueError::for_wiredtiger( result))
  }
}
